snapcraft/internal/lifecycle.py

```diff
@@ -112,54 +112,65 @@ def __init__(self, config, project_options):
         self.config = config
         self.project_options = project_options
         self.parts_config = config.parts
+        self._steps_run = self._init_run_states()
 
-    def run(self, step, part_names=None, recursed=False):
+    def _init_run_states(self):
+        steps_run = {}
+
+        for part in self.config.all_parts:
+            steps_run[part.name] = set()
+            for step in common.COMMAND_ORDER:
+                if part.is_dirty(step):
+                    self._handle_dirty(part, step)
+                elif not (part.should_step_run(step)):
+                    steps_run[part.name].add(step)
+                    part.notify_part_progress('Skipping {}'.format(step),
+                                              '(already ran)')
+
+        return steps_run
+
+    def run(self, step, part_names=None):
         if part_names:
             self.parts_config.validate(part_names)
-            parts = {p for p in self.config.all_parts if p.name in part_names}
+            parts = [p for p in self.config.all_parts if p.name in part_names]
```

Comment:
Typically I see you going from array to set, not the other way. Is there a reason for it?